scipy/optimize/tests/test_optimize.py

```diff
@@ -1259,6 +1259,26 @@ def dfunc(z):
             assert func(sol1.x) < func(sol2.x), \
                    f"{method}: {func(sol1.x)} vs. {func(sol2.x)}"
 
+    @pytest.mark.parametrize(
+        'bounds', [None, [[0.0, 0.0], [-np.inf, +np.inf], [-np.inf, +np.inf]]],
+    )
+    @pytest.mark.parametrize('method', ['l-bfgs-b'])
+    def test_minimize_callback_result(self, method, bounds):
+        """Check if `OptimizeResult` is passed to the callback function.
+
+        The issue related to fixed bounds (gh-21537) is also checked.
+        """
+        def callback(intermediate_result):
+            assert isinstance(intermediate_result, optimize.OptimizeResult)
+
+        res = optimize.minimize(self.func, np.zeros(3), method=method,
+                                bounds=bounds, callback=callback)
+
+        if bounds is not None:
+            for i in range(3):
+                assert bounds[i][0] <= res.x[i]  # check lower bounds
+                assert bounds[i][1] >= res.x[i]  # check upper bounds
+
```

Comment:
Can we do something like,

```
assert (bounds[:, 0] <= res.x).all()
assert (bounds[:, 1] >= res.x).all()
```

Note that `bounds` is kept as a numpy array. I tried one example and it works,

```python
>>> import numpy as np
>>> bounds = [[0.0, 0.0], [-np.inf, +np.inf], [-np.inf, +np.inf]]
>>> bounds = np.array(bounds)
>>> bounds
array([[  0.,   0.],
       [-inf,  inf],
       [-inf,  inf]])
>>> x = [1, 2, 3]
>>> x = np.array([1, 2, 3])
>>> x
array([1, 2, 3])
>>> bounds[:, 0] >= x
array([False, False, False])
>>> bounds[:, 1] <= x
array([ True, False, False])
```

Suggesting to use NumPy API over using Pythonic loops. `res.x` is anyways a NumPy array and we can make `bounds` a NumPy array so we can avoid the loop easily.